File: src/main/java/tinydiff/Diff.java
package tinydiff;

/**
 * Diff
 * @author Kazuhiko Arase
 */
public final class Diff implements IDiff {

  private DiffTargetWrapper wrapper = null;

  private IDiff diff = null;

  public Diff() {
  }

  public int onp(final IDiffComparable target) {
    wrapper = new DiffTargetWrapper(target);
    diff = new DiffImpl();
    return diff.onp(wrapper);
  }

  public void trace(final IPathTracer tracer) {
    if (diff == null) {
      throw new IllegalStateException("onp before trace.");
    }
    diff.trace(new IPathTracer() {
      public void trace(final int deltaX, final int deltaY) {
        if (wrapper.isReverse() ) {
          tracer.trace(deltaY, deltaX);
        } else {
          tracer.trace(deltaX, deltaY);
        }
      }
    } );
  }

  private static final class DiffTargetWrapper
  implements IDiffComparable {
    private final IDiffComparable target;
    private final boolean reverse; 
    public DiffTargetWrapper(final IDiffComparable target) {
      this.target = target;
      this.reverse = target.getM() > target.getN();
    }
    public boolean isReverse() {
      return reverse;
    }
    public boolean equals(final int x, final int y) {
      return reverse? target.equals(y, x) : target.equals(x, y);
    }
    public int getM() {
      return reverse? target.getN() : target.getM();
    }
    public int getN() {
      return reverse? target.getM() : target.getN();
    }
  }
}

File: src/main/ts/tinydiff/Diff.ts
namespace tinydiff {

  interface IDiffTargetWrapper extends IDiffComparable {
    isReverse() : boolean;
  }

  /**
   * createDiff
   * @author Kazuhiko Arase
   */
  export function createDiff() : IDiff {

    var wrapper : IDiffTargetWrapper = null;

    var diff : IDiff = null;

    return {
      onp(target) {
        wrapper = createDiffTargetWrapper(target);
        diff = createDiffImpl();
        return diff.onp(wrapper);
      },
      trace(tracer) {
        if (diff == null) {
          throw 'onp before trace.';
        }
        diff.trace({
          trace(deltaX, deltaY) {
            if (wrapper.isReverse() ) {
              tracer.trace(deltaY, deltaX);
            } else {
              tracer.trace(deltaX, deltaY);
            }
          }
        } );
      }
      }

    function createDiffTargetWrapper(
      target : IDiffComparable
    ) : IDiffTargetWrapper {
      var reverse = target.getM() > target.getN();
      return {
        equals : (x, y) =>
          reverse? target.equals(y, x) : target.equals(x, y),
        getM : () => reverse? target.getN() : target.getM(),
        getN : () => reverse? target.getM() : target.getN(),
        isReverse : () => reverse
      }
    };
  }
}

declare var exports : any;
declare var module : any;

!function(tinydiff : any) {
  if (typeof exports === 'object') {
    module.exports = tinydiff;
  }
}(tinydiff);
